packages/react/src/utils.ts

```diff
@@ -73,4 +74,35 @@ const useSize = (target: React.RefObject<HTMLDivElement>) => {
   return size;
 };
 
-export { mergeProps, LKComponentAttributes, useObservableState, cloneSingleChild, useSize };
+export type TokenizeGrammar = { [type: string]: RegExp };
+
+export function tokenize(input: string, grammar: TokenizeGrammar) {
+  const matches = Object.entries(grammar)
+    .map(([type, rx], weight) =>
+      Array.from(input.matchAll(rx)).map(({ index, 0: content }) => ({
+        type,
+        weight,
+        content,
+        index: index ?? 0,
+      })),
+    )
+    .flat()
+    .sort((a, b) => {
+      const d = a.index - b.index;
+      return d !== 0 ? d : a.weight - b.weight;
+    })
+    .filter(({ index }, i, arr) => {
+      if (i === 0) return true;
+      const prev = arr[i - 1];
+      return prev.index + prev.content.length <= index;
+    });
+
+  const tokens = [];
+  let pos = 0;
+  for (const { type, content, index } of matches) {
```

Comment:
this is currently only looping over matches, without special handling for the case when there's no match at all. This results in messages that are a string without any matches to be returned as an empty array

Reply:
good catch